fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -3489,6 +3490,66 @@ ACTOR Future<Void> processTransactionStateRequestPart(TransactionStateResolveCon
 
 } // anonymous namespace
 
+//
+// Metrics related to the commit proxy are logged on a five second interval in
+// the `ProxyMetrics` trace. However, it can be hard to determine workload
+// burstiness when looking at such a large time range. This function adds much
+// more frequent logging for certain metrics to provide fine-grained insight
+// into workload patterns. The metrics logged by this function break down into
+// two categories:
+//
+//   * existing counters reported by `ProxyMetrics`
+//   * new counters that are only reported by this function
+//
+// Neither is implemented optimally, but the data collected should be helpful
+// in identifying workload patterns on the server.
+//
+// Metrics reporting by this function can be disabled by setting the
+// `BURSTINESS_METRICS_ENABLED` knob to false. The reporting interval can be
+// adjusted by modifying the knob `BURSTINESS_METRICS_LOG_INTERVAL`.
+//
+ACTOR Future<Void> logDetailedMetrics(ProxyCommitData* commitData) {
+	state double startTime = 0;
+	state int64_t commitBatchInBaseline = 0;
+	state int64_t txnCommitInBaseline = 0;
+	state int64_t mutationsBaseline = 0;
+	state int64_t mutationBytesBaseline = 0;
+
+	loop {
+		if (!SERVER_KNOBS->BURSTINESS_METRICS_ENABLED) {
+			return Void();
+		}
+
+		startTime = now();
+		commitBatchInBaseline = commitData->stats.commitBatchIn.getValue();
+		txnCommitInBaseline = commitData->stats.txnCommitIn.getValue();
+		mutationsBaseline = commitData->stats.mutations.getValue();
+		mutationBytesBaseline = commitData->stats.mutationBytes.getValue();
+
+		wait(delay(SERVER_KNOBS->BURSTINESS_METRICS_LOG_INTERVAL));
+
+		int64_t commitBatchInReal = commitData->stats.commitBatchIn.getValue();
+		int64_t txnCommitInReal = commitData->stats.txnCommitIn.getValue();
+		int64_t mutationsReal = commitData->stats.mutations.getValue();
+		int64_t mutationBytesReal = commitData->stats.mutationBytes.getValue();
+
+		// Don't log anything if all the values match the baseline, or if any
+		// of the counters got reset during the wait interval.
+		if (commitBatchInReal <= commitBatchInBaseline && txnCommitInReal <= txnCommitInBaseline &&
```

Comment:
The comment doesn't match the test, which only bails if _all_ of the counters have been reset, not _any_.

This also adds a new test for zero-change metrics, without explaining why.

Given that the systems we care about process thousands of transactions (and batches) per second, it's probably not worth the trouble omitting zero metrics.

If these are polluting simulation logs (not sure if that's the case, just guessing), then maybe disable this logging by default. That's actually the default we probably want for this.

Reply:
> which only bails if all of the counters have been reset, not any

I think technically this should always be an all or nothing change. One of the counters being reset while the others aren't should probably be an assertion failure or something. But it's an assumption that I will document making.

> If these are polluting simulation logs (not sure if that's the case, just guessing), then maybe disable this logging by default. That's actually the default we probably want for this.

I changed the default knob to false, I think that makes sense. As long as it is relatively easy to turn on for periods, I agree that disabling what is probably going to be a high volume trace is a good idea.